scipy/stats/_qmc.py

```diff
@@ -0,0 +1,1587 @@
+    h_i2 = np.prod(1. + 0.5 * abs(z_ij[i2, :]) - 0.5 * (z_ij[i2, :] ** 2))
+
+    # Eq (25), typo in the article g is missing
+    c_p_i1i1 = ((g_i1 * alpha) / (n ** 2) - 2. * alpha * beta * h_i1 / n)
+    # Eq (26), typo in the article n ** 2
+    c_p_i2i2 = ((g_i2 / ((n ** 2) * alpha)) - (2. * h_i2 / (n * alpha * beta)))
+
+    # Eq (26)
+    sum_ = c_p_i1j - c_i1j + c_p_i2j - c_i2j
+
+    mask = np.ones(n, dtype=bool)
+    mask[[i1, i2]] = False
+    sum_ = sum(sum_[mask])
+
+    disc_ep = (disc + c_p_i1i1 - c_i1i1 + c_p_i2i2 - c_i2i2 + 2 * sum_)
+
+    return disc_ep
+
+
+def primes_from_2_to(n: int) -> List[int]:
+    """Prime numbers from 2 to *n*.
+
+    Parameters
+    ----------
+    n : int
+        Sup bound with ``n >= 6``.
+
+    Returns
+    -------
+    primes : list(int)
+        Primes in ``2 <= p < n``.
+
+    References
+    ----------
+    .. [1] `StackOverflow <https://stackoverflow.com/questions/2068372>`_.
+
+    """
+    sieve = np.ones(n // 3 + (n % 6 == 2), dtype=bool)
+    for i in range(1, int(n ** 0.5) // 3 + 1):
+        k = 3 * i + 1 | 1
+        sieve[k * k // 3::2 * k] = False
+        sieve[k * (k - 2 * (i & 1) + 4) // 3::2 * k] = False
+    return np.r_[2, 3, ((3 * np.nonzero(sieve)[0][1:] + 1) | 1)]
+
+
+def n_primes(n: int) -> List[int]:
+    """List of the n-first prime numbers.
+
+    Parameters
+    ----------
+    n : int
+        Number of prime numbers wanted.
+
+    Returns
+    -------
+    primes : list(int)
+        List of primes.
+
+    """
+    primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59,
+              61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127,
+              131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193,
+              197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269,
+              271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349,
+              353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431,
+              433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503,
+              509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599,
+              601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673,
+              677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761,
+              769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857,
+              859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947,
+              953, 967, 971, 977, 983, 991, 997][:n]
+
+    if len(primes) < n:
+        big_number = 2000
+        while 'Not enough primes':
+            primes = primes_from_2_to(big_number)[:n]
+            if len(primes) == n:
+                break
+            big_number += 1000
+
+    return primes
+
+
+def van_der_corput(
+        n: int, base: int = 2, start_index: int = 0,
+        scramble: bool = False,
+        seed: Optional[int, np.random.RandomState] = None
```

Comment:
If I recall, I copied the seed docstring from somewhere else in scipy. I guess it should be consistent all over the place. Do you have any reference I should look at for this?

Reply:
`_doc_random_state` at the top of `scipy/stats/_multivariate.py` looks decent, except for some of the reST formatting (compare source with http://scipy.github.io/devdocs/generated/scipy.stats.multivariate_normal.html).

The other thing to consider is the default for `seed=None` - why not switch to using `Generator` by default? The time to do that is now, once we have a release with `stats.qmc` in it, it cannot be changed anymore.

Reply:
Sure no problem I guess. So I should do my own `check_random_state`? Something like this? (I just adapted the existing `_lib/_utils/check_random_state`. But I thought scipy still relied on numpy < 1.7.
```python
def check_random_state(seed):
    """Turn seed into a np.random.Generator instance.

    If seed is None (or np.random), return the Generator singleton used
    by np.random.
    If seed is an int, return a new Generator instance seeded with seed.
    If seed is already a Generator instance, return it.
    Otherwise, raise ValueError.
    """
    if seed is None or seed is np.random:
        return np.random.Generator()
    if isinstance(seed, (numbers.Integral, np.integer)):
        return np.random.Generator(seed)
    if isinstance(seed, np.random.Generator):
        return seed

    raise ValueError('%r cannot be used to seed a numpy.random.Generator'
                     ' instance' % seed)

```

Reply:
That looks good. You will need need to add:
```
    if seed is None or seed is np.random:
        if not hasattr(np.random, 'Generator'):
            # This can be removed once numpy 1.16 is dropped
            msg = "NumPy 1.16 doesn't have Generator, use either NumPy >= 1.17 or `seed=np.random.RandomState(seed)`"
            raise ValueError(msg)
        return np.random.default_rng()
```
because we likely will have one more SciPy release before dropping NumPy 1.16.5

Reply:
But then should I modify the one in `_lib/_utils/check_random_state` directly?

Reply:
No, that would have a backwards compatibility impact on other functionality using it unfortunately.

Reply:
I am trying to only use `Generator`. And I have the following:

```python
def check_random_state(seed: Optional[_int, np.random.Generator] = None):
    """Turn `seed` into a `numpy.random.Generator` instance.

    Parameters
    ----------
    seed : {None, int, `numpy.random.Generator`}, optional
        If `seed` is None the `numpy.random.Generator` singleton is used.
        If `seed` is an int, a new ``Generator`` instance is used,
        seeded with `seed`.
        If `seed` is already a ``Generator`` instance then that instance is
        used.

    Returns
    -------
    seed : `numpy.random.Generator`
        Random number generator.

    """
    if seed is None or seed is np.random:
        if not hasattr(np.random, 'Generator'):
            # This can be removed once numpy 1.16 is dropped
            msg = ("NumPy 1.16 doesn't have Generator, use either "
                   "NumPy >= 1.17 or `seed=np.random.RandomState(seed)`")
            raise ValueError(msg)
        return np.random.default_rng()
    if isinstance(seed, (numbers.Integral, np.integer)):
        try:
            return np.random.default_rng(seed)
        except AttributeError:
            return np.random.RandomState(seed)
    if isinstance(seed, np.random.RandomState):
        # This can be removed once numpy 1.16 is dropped
        return seed
    if isinstance(seed, np.random.Generator):
        return seed
    raise ValueError('%r cannot be used to seed a numpy.random.Generator'
                     ' instance' % seed)
```

As the seed changed because we don't use anymore `RandomState`, I have to adapt the [doc]tests. It's ok but I am afraid the tests won't pass on Numpy 1.16.5. Is this a problem if I mark all tests with xpass? Or shall I try to seed everything with `RandomState`? In that case sticking with the current `check_random_state` might be more convenient as all tests are already good.

In any case the doc can only talk about `Generator` to call the user to use this. The implementation would still work on 1.16.5 as further in the code there are some try/except to fallback on `RandomState`.

Reply:
I am seeing in some tests that I could mark with something like `@check_version(numpy, '0.17')`

Reply:
> I am trying to only use `Generator`.

This seems fine to me. It may annoy some users who use the legacy `RandomState` out of habit, but I think this is what we'd like - use `Generator` for new code, and `RandomState` only for tests because it guarantees exact reproducibility over releases.

> `if seed is None or seed is np.random:`

Can we drop `np.random` here? There's no reason to let people pass in the module for new code.

> `except AttributeError:`

That would need a warning, or just error out. I'd probably prefer the latter, for consistency with the default of no seed.

>     `if isinstance(seed, np.random.RandomState):`
>        `# This can be removed once numpy 1.16 is dropped`

No, we can't drop that, that'd be a backwards compat break.

Overall this seems a little inconsistent still. I think this is what we'd want:
1. Accept both `Generator` and `RandomState` instances, now and in the future
2. If no seed or integer seed is given, use `default_rng`
3. If (2) and numpy 1.16.5, raise an exception.

Reply:
ok then if I got it right, the following should do it:

```python
def check_random_state(seed: Optional[_int, np.random.Generator] = None):
    """Turn `seed` into a `numpy.random.Generator` instance.

    Parameters
    ----------
    seed : {None, int, `numpy.random.Generator`}, optional
        If `seed` is None the `numpy.random.Generator` singleton is used.
        If `seed` is an int, a new ``Generator`` instance is used,
        seeded with `seed`.
        If `seed` is already a ``Generator`` instance then that instance is
        used.

    Returns
    -------
    seed : `numpy.random.Generator`
        Random number generator.

    """
    if seed is None or isinstance(seed, (numbers.Integral, np.integer)):
        if not hasattr(np.random, 'Generator'):
            # This can be removed once numpy 1.16 is dropped
            msg = ("NumPy 1.16 doesn't have Generator, use either "
                   "NumPy >= 1.17 or `seed=np.random.RandomState(seed)`")
            raise ValueError(msg)
        return np.random.default_rng(seed)
    elif isinstance(seed, (np.random.Generator, np.random.RandomState)):
        return seed
    else:
        raise ValueError('%r cannot be used to seed a numpy.random.Generator'
                         ' instance' % seed)
```

Reply:
Yep that looks good to me.

Reply:
Ok thanks. Then I will use this one and, as you said, pass a `RandomState` in the tests. Still I suggest to not talk about `RandomState` in the docstrings of the classes themselves to promote the usage of `Generator`.

Reply:
agreed